syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.env";
option java_outer_classname = "WorkerProto";
option objc_class_prefix = "HLW";

package worker;

// The greeting service definition.
service Worker {
  rpc ServiceReady (ServiceReadyReq) returns (ServiceReadyResp) {}
  rpc Start (stream StartReq) returns (StartResp) {}
  rpc GetGradients (GetGradientsReq) returns (stream GetGradientsResp) {}
  rpc ReplicateModel (stream ReplicateModelReq) returns (ReplicateModelResp) {}
  rpc Terminate (TerminateReq) returns (TerminateResp) {}
}

message ServiceReadyReq {
  string requestId = 1;
}

message ServiceReadyResp {
  bool ready = 1;
  string requestId = 2;
}

message StartMeta {
  string filename = 1;
  string trainerType = 2;
  string bufferType = 3;
  uint64 bufferSize = 4;
  int64 maxEpisode = 5;
  int64 maxEpisodeStep = 6;
  string actionTransform = 7;
  uint64 version = 8;
  string requestId = 9;
}

message StartReq {
  oneof request {
    StartMeta meta = 1;
    bytes chunkData = 2;
  }
}

message StartResp {
  bool done = 1;
  string err = 2;
  string requestId = 3;
}

message ModelMeta {
  string shapes = 1;
  uint64 version = 2;
  string requestId = 3;
}

message GetGradientsReq {
  string requestId = 1;
}

message GetGradientsResp {
  oneof request {
    ModelMeta meta = 1;
    bytes chunkData = 2;
  }
}

message ReplicateModelReq {
  oneof request {
    ModelMeta meta = 1;
    bytes chunkData = 2;
  }
}

message ReplicateModelResp {
  bool done = 1;
  string err = 2;
  string requestId = 3;
}

message TerminateReq {
  string requestId = 1;
}

message TerminateResp {
  bool done = 1;
  string err = 2;
  string requestId = 3;
}
